File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.js
const input = require("fs")
  .readFileSync(0)
  .toString()
  .trim()
  .split("\n")
  .map((v) => v.trim());

const N = Number(input[0]);
const SCORE = input.slice(1).map((arr) => arr.split(" ").map(Number));

const combinations = (arr, n) => {
  if (n === 1) return arr.map((v) => [v]);

  const result = [];
  arr.forEach((fixed, idx) => {
    const rest = arr.slice(idx + 1);
    const combis = combinations(rest, n - 1);
    const combine = combis.map((v) => [fixed, ...v]);
    result.push(...combine);
  });

  return result;
};

const COMBIS = combinations(
  Array.from({ length: N }, (_, i) => i),
  N / 2
);

const calculate = (arr) => {
  let score = 0;
  for (let i = 0; i < arr.length - 1; i++) {
    for (let j = i + 1; j < arr.length; j++) {
      score += SCORE[arr[i]][arr[j]] + SCORE[arr[j]][arr[i]];
    }
  }
  return score;
};

let minDiff = Infinity;

const backtrack = (depth) => {
  if (depth === Math.floor(COMBIS.length / 2)) return;

  const T1 = COMBIS[depth];
  const T2 = COMBIS[COMBIS.length - 1 - depth];

  const S1 = calculate(T1);
  const S2 = calculate(T2);

  const diff = Math.abs(S1 - S2);

  if (diff === 0) {
    minDiff = diff;
    return;
  }

  minDiff = Math.min(diff, minDiff);
  backtrack(depth + 1);
};

backtrack(0);

console.log(minDiff);

File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.py
import sys
input = sys.stdin.readline

from itertools import combinations

N = int(input())
SCORE = [list(map(int, input().split())) for _ in range(N)]

def calculate(team):
    score = 0
    for i in range(len(team) - 1):
        for j in range(i + 1, len(team)):
            score += SCORE[team[i]][team[j]] + SCORE[team[j]][team[i]]
    return score

COMBIS = list(combinations(range(N), N // 2))
min_diff = float('inf')

for i in range(len(COMBIS) // 2):  # 백트래킹 제거, 단순 반복문으로 변경
    t1, t2 = COMBIS[i], COMBIS[-(i + 1)]
    s1, s2 = calculate(t1), calculate(t2)
    
    min_diff = min(abs(s1 - s2), min_diff)
    
    if min_diff == 0:
        break  # 0이면 더 이상 탐색할 필요 없음

print(min_diff)